plugins/task-plugin/src/machine/machines-picker.ts

```diff
@@ -29,24 +29,75 @@ export class MachinesPicker {
      */
     async pick(containerNames?: string[]): Promise<string> {
         if (!containerNames) {
-            containerNames = await this.cheWorkspaceClient.getContainersNames();
+            return this.pickContainerFromClient();
         }
 
-        if (containerNames.length === 1) {
-            return Promise.resolve(containerNames[0]);
+        return this.pickContainerFromStringList(containerNames);
+    }
+
+    private pickContainerFromStringList(containers: string[]): Promise<string> {
+        if (containers.length === 1) {
+            return Promise.resolve(containers[0]);
         }
-        return this.showMachineQuickPick(containerNames);
+        return this.showQuickPick(containers);
     }
 
-    private showMachineQuickPick(items: string[]): Promise<string> {
+    private showQuickPick(items: string[]): Promise<string>;
+    private showQuickPick(items: QuickPickItem[]): Promise<string>;
+    private showQuickPick(items: string[] | QuickPickItem[]): Promise<string> {
         return new Promise<string>(resolve => {
-
-            const options = { placeHolder: CONTAINERS_PLACE_HOLDER } as theia.QuickPickOptions;
+            const options = { placeHolder: CONTAINERS_PLACE_HOLDER, ignoreFocusOut: false } as QuickPickOptions;
             options.onDidSelectItem = (item => {
                 const machineName = typeof item === 'string' ? item : item.label;
                 resolve(machineName);
             });
-            theia.window.showQuickPick(items, options);
+
+            if (items.length === 0) {
+                return;
+            }
+
+            if (typeof items[0] === 'string') {
+                window.showQuickPick(items as string[], options);
+            } else {
+                window.showQuickPick(items as QuickPickItem[], options);
+            }
         });
     }
+
+    private async pickContainerFromClient(): Promise<string> {
+        const containers = await this.cheWorkspaceClient.getContainers();
+
+        if (containers.length === 1) {
+            return Promise.resolve(containers[0].name);
+        }
+
+        const items = this.mapContainersToFilteredQuickPickItems(containers);
+
+        return this.showQuickPick(items);
+    }
+
+    private mapContainersToFilteredQuickPickItems(containers: WorkspaceContainer[]): QuickPickItem[] {
+        const items: QuickPickItem[] = [];
+
+        const devContainers = containers.filter((container: WorkspaceContainer) => container.attributes
+            && (!container.attributes[CONTAINER_SOURCE_ATTRIBUTE] || container.attributes[CONTAINER_SOURCE_ATTRIBUTE] === RECIPE_CONTAINER_SOURCE));
+        const toolingContainers = containers.filter((container: WorkspaceContainer) => !(container.attributes
+            && (!container.attributes[CONTAINER_SOURCE_ATTRIBUTE] || container.attributes[CONTAINER_SOURCE_ATTRIBUTE] === RECIPE_CONTAINER_SOURCE)));
```

Comment:
If I'm not mistaken you use 

`container.attributes
            && 
(!container.attributes[CONTAINER_SOURCE_ATTRIBUTE] || container.attributes[CONTAINER_SOURCE_ATTRIBUTE] === RECIPE_CONTAINER_SOURCE)`

condition to check is it `devContainer` and the same condition with inversion to check is it `toolingContainer`

Does it make sense to create method  `isDevContainer()` to provide more readable filter?
```
const devContainers = containers.filter(container => this.isDevContainer(container));
const toolingContainers = containers.filter(container => !this.isDevContainer(container));
```
wdyt?

Reply:
fixed